### DISCUSSION TITLE
Corpus submission

### DISCUSSION THREAD

**Discussion by addreas** [cue-unity/unity#72] [General]
Just a quick question about what kinds of repositories would be considered interesting/relevant as part of the corpus. 

I've ended up with a CUE repository for my personal Kubernetes setup [here](https://github.com/addreas/homelab), however I don't think it would be too much fun keeping a test/golden output of everything up to date. So in that case it would just be a single `cue vet` as the test which might not be interesting enough. Being just my home playground it isn't really that big either.

On a larger scale there is the option to create something less real, perhaps a larger version of the [unity-example](https://github.com/cue-sh/unity-example), but still following a similar structure to what can be found "in the wild". It occurred to me that something like that could even primarily be a collection of CUE examples/patterns, that just happens to be a good test as well.

**Comment by myitcv** [comment]
Hi @addreas

> Just a quick question about what kinds of repositories would be considered interesting/relevant as part of the corpus.
> 
> I've ended up with a CUE repository for my personal Kubernetes setup [here](https://github.com/addreas/homelab), however I don't think it would be too much fun keeping a test/golden output of everything up to date. So in that case it would just be a single `cue vet` as the test which might not be interesting enough. Being just my home playground it isn't really that big either.

Great question, and thanks for asking here. Truth is, we'd be happy for anything and everything to be part of the corpus, because it simply serves to give us more coverage, and that has only been a good thing thus far as far as smoothing out our releases is concerned. 

On your point about keeping a test/golden output up to date; strictly speaking there shouldn't be anything to do. An entry in the corpus is a `git` submodule, and so will only advance when told to do so. Thus, the only thing that might break, all else being equal, is a change/fix/bug in a later CUE release as we are testing things. That might well happen, but because we're on the road to `v1.0.0` at which point there will be no breaking changes, save for critical bug fixes (a la the [Go Compatibility Promise](https://golang.org/doc/go1compat)), it is becoming less and less common. As and when it does happen, having the corpus "break" is actually a good way of us testing that our proposed automated fixes/instructions actually work _ahead_ of users having to try them out. 

Note also the `--update` flag which helps automate the updating of golden files.

We also have plans for https://github.com/cue-sh/unity/issues/39, which will automate the process of "advancing" the corpus `git` submodules. A further enhancement on that point would be to say that if we know a given corpus entry does successfully work according to the tests for a later version of CUE than that specified in the manifest, then we could flag as much to the project.

> On a larger scale there is the option to create something less real, perhaps a larger version of the [unity-example](https://github.com/cue-sh/unity-example), but still following a similar structure to what can be found "in the wild". It occurred to me that something like that could even primarily be a collection of CUE examples/patterns, that just happens to be a good test as well.

This is a great idea, and ties in very much with our idea of improving documentation. I developed https://play-with-go.dev/ in conjunction with Google for the Go project. Whilst Play with Go is still in its early days, its precursor served as an excellent set of integration tests whilst Go modules were in _their_ early days. The same will very much be true of "Play with CUE" when we come to get that up and running.

TL;DR - please feel free to suggest anything for the corpus.

**Reply by addreas** [reply]
Threw together a PR, so you could take a look if you think it seems worth adding. Wasn't sure if using the overlays folder or adding tests to the source repository was preferable, so I did both.

**Reply by myitcv** [reply]
Merged!
